subsys/net/lib/wifi_credentials/wifi_credentials.c

```diff
@@ -0,0 +1,538 @@
+	k_mutex_unlock(&wifi_credentials_mutex);
+	return ret;
+}
+
+int wifi_credentials_get_by_ssid_personal(
+	const char *ssid,
+	size_t ssid_len,
+	enum wifi_security_type *type,
+	uint8_t *bssid_buf,
+	size_t bssid_buf_len,
+	char *password_buf,
+	size_t password_buf_len,
+	uint32_t *flags
+)
+{
+	int ret = 0;
+	uint8_t buf[ENTRY_MAX_LEN] = {0};
+
+	if (ssid == NULL || ssid_len > WIFI_SSID_MAX_LEN || ssid_len == 0) {
+		LOG_ERR("SSID has invalid format");
+		return -EINVAL;
+	}
+	if (bssid_buf_len != WIFI_MAC_ADDR_LEN || bssid_buf == NULL) {
+		LOG_ERR("BSSID buffer needs to be provided");
+		return -EINVAL;
+	}
+	if (password_buf == NULL ||
+	    password_buf_len > WIFI_CREDENTIALS_MAX_PASSWORD_LEN ||
+	    password_buf_len == 0) {
+		LOG_ERR("password buffer needs to be provided");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&wifi_credentials_mutex, K_FOREVER);
+
+	/* load entry */
+	ret = wifi_credentials_get_by_ssid_personal_raw(ssid, ssid_len,
+							(struct wifi_credentials_personal *)buf);
+
+	if (ret) {
+		goto exit;
+	}
+
+	/* unpack entry*/
+	struct wifi_credentials_header *header = (struct wifi_credentials_header *)buf;
+
+	*type = header->type;
+	*flags = header->flags;
+
+	if (header->flags & WIFI_CREDENTIALS_FLAG_BSSID) {
+		memcpy(bssid_buf, header->bssid, WIFI_MAC_ADDR_LEN);
+	}
+
+	switch (header->type) {
+	case WIFI_SECURITY_TYPE_NONE:
+		break;
+	case WIFI_SECURITY_TYPE_PSK:
+	case WIFI_SECURITY_TYPE_PSK_SHA256:
+	case WIFI_SECURITY_TYPE_SAE:
+	{
+		struct wifi_credentials_personal *header_personal =
+			(struct wifi_credentials_personal *)buf;
+
+		memcpy(password_buf, header_personal->password, WIFI_CREDENTIALS_MAX_PASSWORD_LEN);
+		break;
+	}
+	default:
+		LOG_ERR("this code section should be unreachable");
+		ret = -EPROTO;
+	}
+exit:
+	k_mutex_unlock(&wifi_credentials_mutex);
+	return ret;
+}
+
+int wifi_credentials_delete_by_ssid(const char *ssid, size_t ssid_len)
+{
+	int ret = 0;
+
+	if (ssid == NULL || ssid_len > WIFI_SSID_MAX_LEN || ssid_len == 0) {
+		LOG_ERR("SSID has invalid format");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&wifi_credentials_mutex, K_FOREVER);
+	ssize_t idx = lookup_idx(ssid);
+
+	if (idx == -1) {
+		LOG_DBG("entry to delete was not found");
+		goto exit;
+	}
+
+	memset(configured_ssids[idx], 0, WIFI_SSID_MAX_LEN);
+
+	ret = wifi_credentials_delete_entry(idx);
+
+exit:
+	k_mutex_unlock(&wifi_credentials_mutex);
+	return ret;
+}
+
+int wifi_credentials_list_ssids(char *ssid_buf[], size_t ssid_buf_len)
+{
+	int ret = 0;
+
+	k_mutex_lock(&wifi_credentials_mutex, K_FOREVER);
+	if (ssid_buf_len < CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES) {
+		ret = -EINVAL;
+		goto exit;
+	}
+	size_t j = 0;
+
+	for (size_t i = 0; i < CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES; ++i) {
+		if (strlen(configured_ssids[i]) > 0) {
+			memcpy(ssid_buf[j], configured_ssids[i], WIFI_SSID_MAX_LEN);
+			j += 1;
+		}
+	}
+exit:
+	k_mutex_unlock(&wifi_credentials_mutex);
+	return ret;
+}
+
+#if defined(CONFIG_WIFI_CREDENTIALS_BACKEND_SETTINGS)
```

Comment:
It might be a good idea to put these backends into separate C files

Alternatively, interlace the definitions for each backend (surround each definition pair with ifdef, rather than do each backend in two large separate blocks)